### PR TITLE
[FLINK-19680][checkpointing] Announce timeoutable CheckpointBarriers

### PR DESCRIPTION
**Pull request by pnowojski**
This PR adds `CheckpointBarrierAnnouncement` messages 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 1]
```diff
@@ -437,19 +440,20 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
...
 				} else {
-					receivedBuffers.add(new SequenceBuffer(buffer, sequenceNumber));
+					receivedBuffers.add(sequenceBuffer);
 					channelStatePersister.maybePersist(buffer);
-				}
 
+					if (dataType.requiresAnnouncement()) {
+						checkPriorityXorAnnouncement(buffer);
+						firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
+					}
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:455]:
```
if (dataType.hasPriority() || dataType.requiresAnnouncement()) {
  firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
}
if (!dataType.hasPriority()) {
  receivedBuffers.add(sequenceBuffer);
  channelStatePersister.maybePersist(buffer);
}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrierAnnouncement.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.flink.runtime.event.RuntimeEvent;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * {@link CheckpointBarrierAnnouncement} is announcing presence or receiving of a {@link CheckpointBarrier}.
+ * That {@link #announcedBarrier} is identified by it's sequence number.
+ */
+public class CheckpointBarrierAnnouncement extends RuntimeEvent {
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrierAnnouncement.java:32]: I was hoping that we could make the announcement mechanism a bit independent of checkpoints. I don't see how we can generalize DataType, but I could imagine having a generic `AnnouncementEvent`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java [hunk 3]
```diff
@@ -292,21 +313,37 @@ public boolean isEvent() {
 			return isEvent;
 		}
 
+		public boolean isBlockingUpstream() {
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:316]: Why did you move this method?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java [hunk 4]
```diff
@@ -247,41 +247,62 @@
...
 		 */
-		ALIGNED_CHECKPOINT_BARRIER(false, true, true, false);
+		ALIGNED_CHECKPOINT_BARRIER(false, true, true, false, false),
+
+		/**
+		 * {@link #TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER} indicates that this buffer represents a
+		 * serialized checkpoint barrier of aligned exactly-once checkpoint mode, that can be time-out'ed
+		 * to an unaligned checkpoint barrier.
+		 */
+		TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER(false, true, true, false, true);
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:280]: As a set before, I'm not a huge fan of this very specific `DataType`s. (Why don't we have an `DataType` just for `EndOfPartitionEvent`s as well?) I also think that `ALIGNED_CHECKPOINT_BARRIER` is on a different level than all other types.

One option that I could see is to just have `ANNOUNCED_EVENT_BUFFER(false, true, true, false, true)`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/SingleCheckpointBarrierHandler.java [hunk 5]
```diff
@@ -147,6 +148,13 @@ public void processBarrier(CheckpointBarrier barrier, InputChannelInfo channelIn
 		}
 	}
 
+	@Override
+	public void processBarrierAnnouncement(
+			CheckpointBarrierAnnouncement barrierAnnouncement,
+			InputChannelInfo channelInfo) throws IOException {
+		// Ignore for now.
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/SingleCheckpointBarrierHandler.java:155]: Add TODO instead?

**Review by AHeise — COMMENTED** [review]
In general the approach and implementation look good. I have two concerns:
- I don't see the benefit to add alignment timeout to `CheckpointOptions`. I'd rather have it on task level pass to `Aligner`.
- I think that these very specific `DataType`s will cause an inflation. I'd like to keep them more high level and only interpret the event type with the payload.

**Comment by pnowojski** [comment]
> I don't see the benefit to add alignment timeout to CheckpointOptions. I'd rather have it on task level pass to Aligner.

if not `alignmentTimeout`, I would still need to add some flag like `isTimeoutable` or `requiresAnnouncement` to the `CheckpointOptions`, to distinguish in the `DataType#getDataType` between:
- aligned non-timeoutable checkpoint/savepoint
- aligned timeoutable checkpoint
- unaligned checkpoint

if we already need this bit of information, why not use it pass the timeout itself?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java [hunk 6]
```diff
@@ -47,29 +50,56 @@
 
 	private final boolean isUnalignedCheckpoint;
 
+	private final long alignmentTimeout;
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java:53]: No, `alignmentTimeout` is a global timeout, measured against the `CheckpointBarrier#getTimestamp`, so since the creation of the `CheckpointBarrier` on the `CheckpointCoordinator`. If `CheckpointCoordinator` has timed out on an upstream task, it will be timed out for any downstream operator as well.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 1 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:455]: This is incorrect, it would have to be:
```
				if (dataType.requiresAnnouncement()) {
					firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
				}
				if (dataType.hasPriority()) {
					firstPriorityEvent = addPriorityBuffer(sequenceBuffer);
				}
				else {
					receivedBuffers.add(sequenceBuffer);
					channelStatePersister.maybePersist(buffer);
				}
```
but that would suggest priority event can be announced, hence I moved 
```
				if (dataType.requiresAnnouncement()) {
					firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
				}
```
into the non priority branch.

**Comment by pnowojski** [comment]
@AHeise, I have replaced `CheckpointBarrierAnnouncement` with `EventAnnouncement` as you were suggesting. Regarding the `DataType.TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER` being inappropriate, it has the same pre-existing problem as  `ALIGNED_CHECKPOINT_BARRIER`, since both require `isBlockingUpstream` flag.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 1 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: Missed the `announce` part.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrierAnnouncement.java (hunk 2 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrierAnnouncement.java]: Okay let's go with specialized and generalize when a second use case pops up.

**Review by AHeise — APPROVED** [review]
LGTM % test failure.

**Comment by pnowojski** [comment]
Azure green, merging.
